app/code/Magento/Theme/Block/Html/Topmenu.php

```diff
@@ -213,18 +214,19 @@ protected function _getHtml(
         $parentLevel = $menuTree->getLevel();
         $childLevel = $parentLevel === null ? 0 : $parentLevel + 1;
 
+        if ($childLevel === 0) {
+            $this->removeOrphans($children);
+        }
+
         $counter = 1;
         $itemPosition = 1;
         $childrenCount = $children->count();
 
         $parentPositionClass = $menuTree->getPositionClass();
         $itemPositionClassPrefix = $parentPositionClass ? $parentPositionClass . '-' : 'nav-';
 
-        /** @var \Magento\Framework\Data\Tree\Node $child */
+        /** @var Node $child */
         foreach ($children as $child) {
-            if ($childLevel === 0 && $child->getData('is_parent_active') === false) {
-                continue;
-            }
             $child->setLevel($childLevel);
             $child->setIsFirst($counter == 1);
             $child->setIsLast($counter == $childrenCount);
```

Comment:
Approach looks generally good to me but I don't the idea of separate filtering loop.

I believe instead of removing `if` and introducing filtering we need just to track last item properly.

For example, remove `$childrenCount` variable, declare `$lastChild = null` before loop, do 
```
$child->setIsLast(false);
$lastChild = $child;
```
in loop and
```
if ($lastChild) {
    $lastChild->setIsLast(true);
}
```
after it.

As always, please squash changes into single commit and force push after changes are performed.

Reply:
@orlangur thank you for your feedback.

Your approach was also the first thing that I wanted to do, but that does not work because the `$html` with all the html for the `$schildren` is generated in the `foreach` loop.  

If you do the `setIsLast()` after the loop the class `last` is not added to the last active menu item (which was the whole issue for this commit ;) )

_I squashed the commits :+1: _

Reply:
1. Please remove `$itemPosition` as it seems to be unused
1. Track `$lastChild` as requested earlier
1. Split one loop into two: in first we process children, in second generate actual HTML with concatenation 
1. Feel free to introduce new private methods to get rid of `@SuppressWarnings(PHPMD.CyclomaticComplexity)` and `@SuppressWarnings(PHPMD.NPathComplexity)`